File: Development/Web UI/sqldm-d3zk/src/com/idera/sqldm/d3zk/chart/PieChart.java
package com.idera.sqldm.d3zk.chart;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import java.util.regex.Matcher;

import org.zkoss.json.JSONObject;
import org.zkoss.json.JSONValue;
import org.zkoss.zk.ui.sys.ContentRenderer;
import org.zkoss.zul.PieModel;
import org.zkoss.zul.SimplePieModel;

public class PieChart extends AbstractChart {

	static final long serialVersionUID = 1L;

	protected PieModel model = new SimplePieModel();

	protected Boolean showSliceLabels = false;

	protected Boolean animateSlices = true;
	protected Integer animationLength = 500;
	protected Integer animationDelay = 50;

	protected Integer innerRadiusPercentage = 0;
	protected Integer outerRadiusPercentage = 85;

	protected String valueFormat = "";

	protected String sortDirection = "asc";

	protected float labelThreshold = 0.05F;
	protected boolean mouseoverEnabled = true;

	protected Map<String, String> colorCodes = new HashMap<String, String>();

	/**
	 * Gets the PieModel containing the data for the chart.
	 *
	 * The model has a Comparable<?> key that should be unique for the category and a
	 * Number for the value.
	 *
	 * The PieModel is a ZK built-in object and the most common implementation is SimplePieModel.
	 *
	 * @return Returns the map containing the data for the chart
	 */
	public PieModel getModel() {
		return model;
	}

	/**
	 * Set the PieModel containing the data for the chart.
	 *
	 * The model has a Comparable<?> key that should be unique for the category and a
	 * Number for the value.
	 *
	 * The PieModel is a ZK built-in object and the most common implementation is SimplePieModel.
	 *
	 * @param model The PieModel to use for the chart data.
	 */
	public void setModel(PieModel model) {
		this.model = model;
	}

	/**
	 * Returns whether or not a label will be displayed on each individual slice of the pie chart.
	 *
	 * If set to true, then a label with either the key or the value for the slice will be rendered.
	 *
	 * The setSliceLabelType(...) method controls whether the key or the value is displayed.
	 *
	 * @return Returns true if the slice labels are shown, false otherwise.
	 */
	public Boolean getShowSliceLabels() {
		return showSliceLabels;
	}

	/**
	 * Sets whether or not a label will be displayed on each individual slice of the pie chart.
	 *
	 * If set to true, then a label with either the key or the value for the slice will be rendered.
	 *
	 * The setSliceLabelType(...) method controls whether the key or the value is displayed.
	 *
	 * @param showSliceLabels True if a label for each slice should be shown, false if not.
	 */
	public void setShowSliceLabels(Boolean showSliceLabels) {
		this.showSliceLabels = showSliceLabels;
	}

	/**
	 * Gets whether or not the slice of the chart will be animated when the chart is
	 * displayed or resized.
	 *
	 * @return Returns true if the slices will be animated when the chart is displayed, false if not.
	 */
	public Boolean getAnimateSlices() {
		return animateSlices;
	}

	/**
	 * Sets whether or not the slices of the chart will be animated when the chart is
	 * displayed or resized.
	 *
	 * @param animateSlices True if the slices will be animated when the chart is displayed, false if not.
	 */
	public void setAnimateSlices(Boolean animateSlices) {
		this.animateSlices = animateSlices;
	}

	/**
	 * Gets the delay between when the chart is first displayed and when the bar animation begins
	 * in milliseconds.
	 *
	 * This field has no effect if animateSlices is set to false.
	 *
	 * @return Returns the delay for when the animation begins in milliseconds
	 */
	public Integer getAnimationDelay() {
		return animationDelay;
	}

	/**
	 * Sets the delay between when the chart is first displayed and when the bar animation begins
	 * in milliseconds.
	 *
	 * This field has no effect if animateSlices is set to false.
	 *
	 * @param animationDelay The number of milliseconds to wait between when the chart is first displayed and when the animation begins
	 */
	public void setAnimationDelay(Integer animationDelay) {
		this.animationDelay = animationDelay;
	}

	/**
	 * Gets the length of the bar animation in milliseconds.
	 *
	 * This field has no effect if animateSlices is set to false.
	 *
	 * @return Returns the length of the bar animation in milliseconds.
	 */
	public Integer getAnimationLength() {
		return animationLength;
	}

	/**
	 * Sets the length of the bar animation in milliseconds.
	 *
	 * This field has no effect if animateSlices is set to false.
	 *
	 * @param animationLength The length of the bar animation in milliseconds.
	 */
	public void setAnimationLength(Integer animationLength) {
		this.animationLength = animationLength;
	}

	/**
	 * Gets the inner radius percentage as an integer between 0 and 100.
	 *
	 * This value indicates whether or not the chart will be displayed as a ring instead of a pie.
	 *
	 * It is calculated as a percentage of the outer radius of the graph.
	 *
	 * If the graph has a 1000px radius, and the innerRadiusPercentage is set to 85, then the ring
	 * will be formed by removing an inner circle with a radius of 850px from the pie.
	 *
	 * @return Returns the inner radius percentage as an integer between 0 and 100
	 */
	public Integer getInnerRadiusPercentage() {
		return innerRadiusPercentage;
	}

	/**
	 * Sets the inner radius percentage as an integer between 0 and 100.
	 *
	 * This value indicates whether or not the chart will be displayed as a ring instead of a pie.
	 *
	 * It is calculated as a percentage of the outer radius of the graph.
	 *
	 * If the graph has a 1000px radius, and the innerRadiusPercentage is set to 85, then the ring
	 * will be formed by removing an inner circle with a radius of 850px from the pie.
	 *
	 * Specified values larger than 100 will set the percentage to 100, values less than 0 will set the percentage to 0.
	 *
	 * @param innerRadiusPercentage The inner radius percentage as an integer between 0 and 100
	 */
	public void setInnerRadiusPercentage(Integer innerRadiusPercentage) {
		if( innerRadiusPercentage > 100 ) innerRadiusPercentage = 100;
		if( innerRadiusPercentage < 0 ) innerRadiusPercentage = 0;
		this.innerRadiusPercentage = innerRadiusPercentage;
	}

	/**
	 * Gets the outer radius percentage as an integer between 0 and 100.
	 *
	 * This value indicates how much of the available space of the component the chart itself will take up.
	 *
	 * It is calculated based on the width or height of the component, whichever is smaller.
	 *
	 * If the component width is 1000, and the outerRadiusPercentage is set to 85, then the graph will be 850px wide.
	 *
	 * @return Returns the outer radius percentage as an integer between 0 and 100
	 */
	public Integer getOuterRadiusPercentage() {
		return outerRadiusPercentage;
	}

	/**
	 * Sets the outer radius percentage as an integer between 0 and 100.
	 *
	 * This value indicates how much of the available space of the component the chart itself will take up.
	 *
	 * It is calculated based on the width or height of the component, whichever is smaller. If the component width is 1000,
	 * and the outerRadiusPercentage is set to 85, then the graph will be 850px wide.
	 *
	 * Specified values larger than 100 will set the percentage to 100, values less than 0 will set the percentage to 0.
	 *
	 * @param outerRadiusPercentage The outer radius percentage as an integer between 0 and 100
	 */
	public void setOuterRadiusPercentage(Integer outerRadiusPercentage) {
		if( outerRadiusPercentage > 100 ) outerRadiusPercentage = 100;
		if( outerRadiusPercentage < 0 ) outerRadiusPercentage = 0;
		this.outerRadiusPercentage = outerRadiusPercentage;
	}

	/**
	 * Gets the format applied to the values of the plot.
	 *
	 * The value is specified as a format string that the d3js library recognizes. This format
	 * is specific to the JS library that the graphs are based on.
	 *
	 * @see <a href="https://github.com/mbostock/d3/wiki/Formatting">https://github.com/mbostock/d3/wiki/Formatting</a>
	 *
	 * The default is an empty string, indicating that no format will be applied.
	 *
	 * This field has no effect if showSliceLabels is false.
	 *
	 * @return Returns the format applied to the values of the plot
	 */
	public String getValueFormat() {
		return valueFormat;
	}

	/**
	 * Sets the format applied to the values of the plot.
	 *
	 * The value is specified as a format string that the d3js library recognizes. This format
	 * is specific to the JS library that the graphs are based on.
	 *
	 * @see <a href="https://github.com/mbostock/d3/wiki/Formatting">https://github.com/mbostock/d3/wiki/Formatting</a>
	 *
	 * The default is an empty string, indicating that no format will be applied.
	 *
	 * This field has no effect if showSliceLabels is false.
	 *
	 * @param valueFormat The format to apply to the values of the plot
	 */
	public void setValueFormat(String valueFormat) {
		this.valueFormat = valueFormat;
	}

	/**
	 * Gets the sort order for the slices in the graph.
	 *
	 * The slices are ordered clockwise starting at the top of the graph.
	 *
	 * @return Returns "asc" if the slices are ordered in ascending order or "dsc" if the slices are ordered in descending order
	 */
	public String getSortDirection() {
		return sortDirection;
	}

	/**
	 * Sets the sort order for the slices in the graph.
	 *
	 * The slices are ordered clockwise starting at the top of the graph.
	 *
	 * @param sortDirection "asc" if the slices should be ordered in ascending order or "dsc" if the slices should be ordered in descending order
	 */
	public void setSortDirection(String sortDirection) {
		this.sortDirection = sortDirection;
	}

	/**
	 * Gets the minimum percentage of the pie that a slice must be to display the slice label.
	 *
	 * Defaults to 0.05 (5%)
	 *
	 * @return Returns the minimum percentage required to show the slice label
	 */
	public float getLabelThreshold() {
		return labelThreshold;
	}

	/**
	 * Gets the minimum percentage of the pie that a slice must be to display the slice label.
	 *
	 * Defaults to 0.05 (5%)
	 *
	 * @param labelThreshold The minimum percentage required to show the slice label
	 */
	public void setLabelThreshold(float labelThreshold) {
		this.labelThreshold = labelThreshold;
	}

	/**
	 * Gets whether or not mouseover events are enabled for the chart slices.
	 *
	 * @return True if mouseover events are enabled, false if not
	 */
	public boolean isMouseoverEnabled() {
		return mouseoverEnabled;
	}

	/**
	 * Sets whether or not mouseover events are enabled for the chart slices.
	 *
	 * @param mouseoverEnabled True if mouseover events should be enabled, false if not
	 */
	public void setMouseoverEnabled(boolean mouseoverEnabled) {
		this.mouseoverEnabled = mouseoverEnabled;
	}

	@Override
	protected void renderProperties(ContentRenderer renderer) throws IOException {
		super.renderProperties(renderer);
		renderer.render("model", toJSON(model));
		renderer.render("showSliceLabels", showSliceLabels);
		renderer.render("animateSlices", animateSlices);
		renderer.render("animationLength", animationLength);
		renderer.render("animationDelay", animationDelay);
		renderer.render("innerRadiusPercentage", innerRadiusPercentage);
		renderer.render("outerRadiusPercentage", outerRadiusPercentage);
		renderer.render("valueFormat", valueFormat);
		renderer.render("sortDirection", sortDirection);
		renderer.render("labelThreshold", labelThreshold);
		renderer.render("mouseoverEnabled", mouseoverEnabled);
		render(renderer, "colorCodes", colorCodes);
	}

	protected static JSONObject toJSON(PieModel model) {

		JSONObject json = new JSONObject();

		for( Comparable<?> category : model.getCategories() ) {

			String newKey = String.valueOf(category);

			newKey = JSONValue.toJSONString(newKey);
			newKey = newKey.substring(1, newKey.length()-1);
			newKey = newKey.replaceAll("[']", Matcher.quoteReplacement("\\'"));

			json.put(newKey, model.getValue(category));
		}

		return json;
	}

	/**
	 * Get the map containing color codes for all labels
	 * 
	 * @return Returns the map containing color codes for all labels
	 */
	public Map<String, String> getColorCodes() {
		return colorCodes;
	}

	/**
	 * Sets the value of map containing color codes
	 * 
	 * @param colorCodes Map of all color codes
	 */
	public void setColorCodes(Map<String, String> colorCodes) {
		this.colorCodes = colorCodes;
	}

}

File: Development/Web UI/sqldm-d3zk/src/web/js/com/idera/sqldm/d3zk/chart/PieChart.js
zk.$package('com.idera.sqldm.d3zk.chart');
com.idera.sqldm.d3zk.chart.PieChart = zk.$extends(com.idera.sqldm.d3zk.chart.AbstractChart, {
	
	_hasRendered: false,
	_showSliceLabels: false,
	_sortDirection: "asc",
	
	_model: {},
	
	_animateSlices: true,
	_animationDelay: 50,
	_animationLength: 500,
	
	_innerRadiusPercentage: 0,
	_outerRadiusPercentage: 85,
	
	_valueFormat: "",
	_labelThreshold: 0.05,
	_mouseoverEnabled: true,
	
	_colorCodes : null,

	getModel: function() { 
		return this._model;
	},
	
	setModel: function(value) { 
		this._model = value;
		if( this._hasRendered ) this.rerenderChart();
	},
	
	getShowSliceLabels: function() { 
		return this._showSliceLabels;
	},
	
	setShowSliceLabels: function(value) { 
		if( this._showSliceLabels != value ) { 
			this._showSliceLabels = value;
			if( this._hasRendered ) this.rerenderChart();
		}
	},
	
	getAnimateSlices: function() { 
		return this._animateSlices;
	},
	
	setAnimateSlices: function(value) { 
		if( value != this._animateSlices ) {
			this._animateSlices = value;
		}
	},
	
	getAnimationDelay: function() {
		return this._animationDelay;
	},
	
	setAnimationDelay: function(value) {
		if( value != this._animationDelay ) {
			this._animationDelay = value;
		}
	},
	
	getAnimationLength: function() {
		return this._animationLength;
	},
	
	setAnimationLength: function(value) {
		if( value != this._animationLength ) {
			this._animationLength = value;
		}
	},	
	
	getInnerRadiusPercentage: function() { 
		return this._innerRadiusPercentage;
	},
	
	setInnerRadiusPercentage: function(value) {
		if( value != this._innerRadiusPercentage ) {
			this._innerRadiusPercentage = value;
			if( this._hasRendered ) this.rerenderChart();
		}
	},

	getOuterRadiusPercentage: function() { 
		return this._outerRadiusPercentage;
	},
	
	setOuterRadiusPercentage: function(value) {
		if( value != this._outerRadiusPercentage ) {
			this._outerRadiusPercentage = value;
			if( this._hasRendered ) this.rerenderChart();
		}
	},
	
	getValueFormat: function() { 
		return this._valueFormat;
	},
	
	setValueFormat: function(value) { 
		if( value != this._valueFormat ) { 
			this._valueFormat = value;
			if( this._hasRendered ) this.rerenderChart();
		}
	},
	
	getSortDirection: function() { 
		return this._sortDirection;
	},
	
	setSortDirection: function(value) {
		if( value != this._sortDirection ) {
			this._sortDirection = value;
			if( this._hasRendered ) this.rerenderChart();
		}
	},
	
	getLabelThreshold: function() { 
		return this._labelThreshold;
	},
	
	setLabelThreshold: function(value) { 
		if( value != this._labelThreshold ) { 
			this._labelThreshold = value;
			if( this._hasRendered ) this.rerenderChart();			
		}
	},
	
	getMouseoverEnabled: function() { 
		return this._mouseoverEnabled;
	},
	
	setMouseoverEnabled: function(value) { 
		if( value != this._mouseoverEnabled ) { 
			this._mouseoverEnabled = value;
			if( this._hasRendered ) this.rerenderChart();
		}
	},
	
	getColorCodes : function() {
		return this._colorCodes;
	},

	setColorCodes : function(value) {
		this._colorCodes = value;
	},

	getZclass: function(){
		var zcls = this._zclass;
		return zcls ? zcls : "z-d3-piechart";
	},
	
	rerenderChart: function() {
		
		var div = document.getElementById(this.uuid);
		if( div == null ) return;
		
		var uuid = this.uuid;
		var svg = d3.select("#" + this.uuid).select("svg");
		if( svg ) svg.remove();
		
		var clientWidth = parseInt(div.clientWidth);
		var clientHeight = parseInt(div.clientHeight);
		
		var width = clientWidth;
		var height = clientHeight;

		var min = Math.min(width, height);
		var outerRadius = (min / 2) * (this._outerRadiusPercentage / 100);
		var innerRadius = outerRadius * (this._innerRadiusPercentage / 100);
		
		var map = d3.map(this._model);
		var entries = map.entries();
		
		var arc = d3.svg.arc()
			.startAngle(function(d){ return d.startAngle; })
			.endAngle(function(d){ return d.endAngle; })		
			.outerRadius(outerRadius)
			.innerRadius(innerRadius);

		var sortAsc = function(a,b) { return a.value < b.value ? -1 : a.value > b.value ? 1 : 0; };
		var sortDsc = function(a,b) { return b.value < a.value ? -1 : b.value > a.value ? 1 : 0; };

		var array_keys = new Array();
		var array_values = new Array();
		var color;
		
		if(this._colorCodes) {
		
			for (var key in this._colorCodes) {
			    array_keys.push(key);
			    array_values.push(this._colorCodes[key]);
			}
			
			color = d3.scale.ordinal()
			 .domain(array_keys)
		 .range(array_values);
		
		} else {
			color = this._getColorPalette();
		}
		
		var pie = d3.layout.pie()
			.sort( this._sortDirection == "asc" ? sortAsc : sortDsc )
			.value(function(d){ return d.value; });

		var svg = d3.select("#" + this.uuid).append("svg")
			.attr("width", clientWidth)
			.attr("height", clientHeight)
			.append("g")
			.attr("transform", "translate(" + (width / 2) + "," + (height / 2) + ")")
			.attr("width", clientWidth)
			.attr("height", clientHeight);

		var g = svg.append("g").attr("class", "arc");

		var paths = g.selectAll("path").data(pie(entries));		
		
		if( this._animateSlices ) { 

			var pieTween = function(d, i) { 
				var i = d3.interpolate({startAngle: 0, endAngle: 0}, {startAngle: d.startAngle, endAngle: d.endAngle});
				return function(t) {
					var b = i(t);
					return arc(b);
				};
			};
			
			paths.enter().append("svg:path")
				.attr("d", arc)
				.attr("fill", function(d, i) { return color(d.data.key); })
				.transition()
				.duration(this._animationLength)
				.delay(this._animationDelay)
				.attrTween("d", pieTween);
			
		}
		else { 
			paths.enter().append("svg:path")
				.attr("d", arc)
				.attr("fill", function(d, i) { return color(d.data.key); });
		}
				
		if( this._showSliceLabels ) { 
			
			var textOffset = 24;
			var labelThreshold = this._labelThreshold; 
			
			var textTween = function(d, i) { 
				
				var a = 0;
				var b = (d.startAngle + d.endAngle - Math.PI)/2;

				var fn = d3.interpolateNumber(a, b);
				return function(t) {
					var val = fn(t);
					return "translate(" + Math.cos(val) * (outerRadius+textOffset) + "," + Math.sin(val) * (outerRadius+textOffset) + ")";
				};				
				
			};
			
			var lines = g.selectAll("line").data(pie(entries));
			
			lines.enter().append("svg:line")
				.attr("x1", 0)
				.attr("x2", 0)
				.attr("y1", -outerRadius-3)
				.attr("y2", -outerRadius-15)
				.attr("class", function(d) { 
					var clazz = "label-tick"; 
					var percent = (d.endAngle - d.startAngle) / (2 * Math.PI);
					if( percent < labelThreshold ) clazz += " hidden";
					return clazz;
				})
				.attr("index", function(d,i) { return i; })
				.attr("transform", function(d) {
					return "rotate(" + (d.startAngle+d.endAngle)/2 * (180/Math.PI) + ")";
				});

			lines.transition()
				.duration(this._animationLength)
				.delay(this._animationDelay)
				.attr("transform", function(d) {
					return "rotate(" + (d.startAngle+d.endAngle)/2 * (180/Math.PI) + ")";
				});

			lines.exit().remove();

			var valueFormat = d3.format(this._valueFormat);
			
			var valueLabels = g.selectAll("text.value").data(pie(entries));

			valueLabels.enter().append("svg:text")
				.attr("class", function(d) { 
					var clazz = "value"; 
					var percent = (d.endAngle - d.startAngle) / (2 * Math.PI);
					if( percent < labelThreshold ) clazz += " hidden";
					return clazz;
				})
				.attr("transform", function(d) {
					return "translate(" + Math.cos(((d.startAngle+d.endAngle - Math.PI)/2)) * (outerRadius+textOffset) + "," + Math.sin((d.startAngle+d.endAngle - Math.PI)/2) * (outerRadius+textOffset) + ")";
				})
				.attr("dy", function(d){
					var isPastHalf = (d.startAngle+d.endAngle)/2 > Math.PI/2 && (d.startAngle+d.endAngle)/2 < Math.PI*1.5;
					return isPastHalf ? 5 : -7;
				})
				.attr("text-anchor", function(d){
					return (d.startAngle+d.endAngle)/2 < Math.PI ? "beginning" : "end";
				})
				.attr("index", function(d,i) { return i; })
				.text(function(d){
					return valueFormat(d.value);
				});

			valueLabels.transition()
				.duration(this._animationLength)
				.delay(this._animationDelay)
				.attrTween("transform", textTween);

			valueLabels.exit().remove();
			
			var nameLabels = g.selectAll("text.units").data(pie(entries));

			nameLabels.enter().append("svg:text")
				.attr("class", function(d) { 
					var clazz = "units"; 
					var percent = (d.endAngle - d.startAngle) / (2 * Math.PI);
					if( percent < labelThreshold ) clazz += " hidden";
					return clazz;
				})
				.attr("transform", function(d) {
					return "translate(" + Math.cos(((d.startAngle+d.endAngle - Math.PI)/2)) * (outerRadius+textOffset) + "," + Math.sin((d.startAngle+d.endAngle - Math.PI)/2) * (outerRadius+textOffset) + ")";
				})
				.attr("dy", function(d){
					var isPastHalf = (d.startAngle+d.endAngle)/2 > Math.PI/2 && (d.startAngle+d.endAngle)/2 < Math.PI*1.5;
					return isPastHalf ? 17 : 5;
				})
				.attr("text-anchor", function(d){
					return (d.startAngle+d.endAngle)/2 < Math.PI ? "beginning" : "end";
				})
				.attr("index", function(d,i) { return i; })
				.text(function(d){
					return d.data.key;
				});

			g.selectAll("text.units").each(function(d, i) {

				var select = d3.select(this);
				if( select.attr("class").indexOf("hidden") != -1 ) return;
				
				var node = select.node();
				if( node == null ) return;
				
				var bbox = node.getBBox();
				var xShift = Math.abs(Math.cos(((d.startAngle+d.endAngle - Math.PI)/2)) * (outerRadius+textOffset));
				
				while( (bbox.width + xShift) > width/2 ) {
					
					var text = d3.select(this).node(); 
					text.textContent = text.textContent.substring(0, text.textContent.length-1) + "...";
					if( text.textContent.length <= 4 ) break;
					
					bbox = d3.select(this).node().getBBox();
					if( (bbox.width + xShift) > width/2 ) text.textContent = text.textContent.substring(0, text.textContent.length-3);
				}
				
			});			
			
			nameLabels.transition()
				.duration(this._animationLength)
				.delay(this._animationDelay)
				.attrTween("transform", textTween);

			nameLabels.exit().remove();			
		
			if( this._showSliceLabels && this._mouseoverEnabled ) { 
	
				paths.on('mouseover', function(d,i) {
					
					d3.select("#" + uuid).selectAll("text.units").each(function(d) {
						
						var label = d3.select(this);
						
						if( label.attr("index") != i ) return;

						label.attr("class", "units highlighted");
						
						var node = label.node();
						if( node == null ) return;
						
						var bbox = node.getBBox();
						var xShift = Math.abs(Math.cos(((d.startAngle+d.endAngle - Math.PI)/2)) * (outerRadius+textOffset));
						
						while( (bbox.width + xShift) > width/2 ) {
							
							var text = d3.select(this).node(); 
							text.textContent = text.textContent.substring(0, text.textContent.length-1) + "...";
							if( text.textContent.length <= 4 ) break;
							
							bbox = d3.select(this).node().getBBox();
							if( (bbox.width + xShift) > width/2 ) text.textContent = text.textContent.substring(0, text.textContent.length-3);
						}						
						
					});

					d3.select("#" + uuid).selectAll("line.label-tick").each(function(d) {
						var label = d3.select(this);
						if( label.attr("index") == i ) label.attr("class", "label-tick highlighted");
					});
					
					d3.select("#" + uuid).selectAll("text.value").each(function(d) {
						var label = d3.select(this);
						if( label.attr("index") == i ) label.attr("class", "value highlighted");
					});
					
				});
				
				paths.on('mouseout', function(d,i) {
					
					d3.select("#" + uuid).selectAll("text.units").each(function(d) {

						var clazz = "units";
						var percent = (d.endAngle - d.startAngle) / (2 * Math.PI);
						if( percent < labelThreshold ) clazz += " hidden";
						
						var label = d3.select(this);
						if( label.attr("index") == i ) label.attr("class", clazz);

					});
					
					d3.select("#" + uuid).selectAll("line.label-tick").each(function(d) {
						
						var clazz = "label-tick";
						var percent = (d.endAngle - d.startAngle) / (2 * Math.PI);
						if( percent < labelThreshold ) clazz += " hidden";
						
						var label = d3.select(this);
						if( label.attr("index") == i ) label.attr("class", clazz);

					});
					
					d3.select("#" + uuid).selectAll("text.value").each(function(d) {
						
						var clazz = "value";
						var percent = (d.endAngle - d.startAngle) / (2 * Math.PI);
						if( percent < labelThreshold ) clazz += " hidden";
						
						var label = d3.select(this);
						if( label.attr("index") == i ) label.attr("class", clazz);

					});  			

				});
	
			}
			
		}
		
	}
});